File: utils/bike_track.py
# -*- coding: utf-8 -*-
"""Сборка GeoJSON трека велотренировки с метаданными точек."""
from __future__ import annotations

import math
from typing import Any


def _arr(values: list[Any]) -> list[Any]:
    return values


def build_enriched_geojson(track_points: list[dict[str, Any]]) -> dict[str, Any] | None:
    """
    track_points: упорядоченные dict с ключами lon, lat, elapsed_sec и опционально
    speed_kmh, cadence, elevation_m, temperature_c, heart_rate, distance_m, power_watts.
    """
    if len(track_points) < 2:
        return None
    coords: list[list[float]] = []
    elapsed: list[int | None] = []
    speed: list[float | None] = []
    cadence: list[float | None] = []
    elevation: list[float | None] = []
    temperature: list[float | None] = []
    heart_rate: list[int | None] = []
    distance_m: list[float | None] = []
    power_watts: list[float | None] = []

    for p in track_points:
        lon, lat = p.get("lon"), p.get("lat")
        if lon is None or lat is None:
            continue
        coords.append([float(lon), float(lat)])
        elapsed.append(
            float(p["elapsed_sec"]) if p.get("elapsed_sec") is not None else None
        )
        speed.append(_float_or_none(p.get("speed_kmh")))
        cadence.append(_float_or_none(p.get("cadence")))
        elevation.append(_float_or_none(p.get("elevation_m")))
        temperature.append(_float_or_none(p.get("temperature_c")))
        heart_rate.append(
            int(p["heart_rate"]) if p.get("heart_rate") is not None else None
        )
        distance_m.append(_float_or_none(p.get("distance_m")))
        power_watts.append(_float_or_none(p.get("power_watts")))

    if len(coords) < 2:
        return None

    return {
        "type": "FeatureCollection",
        "features": [
            {
                "type": "Feature",
                "geometry": {"type": "LineString", "coordinates": coords},
                "properties": {
                    "elapsed_sec": _arr(elapsed),
                    "speed_kmh": _arr(speed),
                    "cadence": _arr(cadence),
                    "elevation_m": _arr(elevation),
                    "temperature_c": _arr(temperature),
                    "heart_rate": _arr(heart_rate),
                    "distance_m": _arr(distance_m),
                    "power_watts": _arr(power_watts),
                },
            }
        ],
    }


def _float_or_none(val: Any) -> float | None:
    if val is None:
        return None
    try:
        return float(val)
    except (TypeError, ValueError):
        return None


def geojson_to_track_points(geo: dict[str, Any]) -> list[dict[str, Any]]:
    """Точки трека с properties (elapsed_sec, speed_kmh, …)."""
    feature = None
    if geo.get("type") == "FeatureCollection" and geo.get("features"):
        feature = geo["features"][0]
    elif geo.get("type") == "Feature":
        feature = geo
    elif geo.get("type") == "LineString":
        feature = {"geometry": geo, "properties": {}}

    if not feature:
        return linestring_to_track_points(geo)

    geometry = feature.get("geometry") or {}
    props = feature.get("properties") or {}
    coords = geometry.get("coordinates") or []
    if geometry.get("type") != "LineString" or len(coords) < 2:
        return linestring_to_track_points(geo)

    elapsed = props.get("elapsed_sec")
    has_props = isinstance(elapsed, list) and len(elapsed) > 0

    out: list[dict[str, Any]] = []
    for i, c in enumerate(coords):
        if not isinstance(c, (list, tuple)) or len(c) < 2:
            continue
        lon, lat = float(c[0]), float(c[1])
        sec = float(elapsed[i]) if has_props and i < len(elapsed) else float(i)
        out.append(
            {
                "lon": lon,
                "lat": lat,
                "elapsed_sec": sec,
                "speed_kmh": _prop_at(props, "speed_kmh", i),
                "cadence": _prop_at(props, "cadence", i),
                "elevation_m": _prop_at(props, "elevation_m", i),
                "temperature_c": _prop_at(props, "temperature_c", i),
                "heart_rate": _prop_int(props, "heart_rate", i),
                "distance_m": _prop_at(props, "distance_m", i),
                "power_watts": _prop_at(props, "power_watts", i),
            }
        )
    return out


def _prop_at(props: dict[str, Any], key: str, i: int) -> float | None:
    arr = props.get(key)
    if not isinstance(arr, list) or i >= len(arr):
        return None
    return _float_or_none(arr[i])


def _prop_int(props: dict[str, Any], key: str, i: int) -> int | None:
    v = _prop_at(props, key, i)
    if v is None:
        return None
    return int(v)


def linestring_to_track_points(geo: dict[str, Any]) -> list[dict[str, Any]]:
    """Из LineString или FeatureCollection извлечь точки (без properties)."""
    coords: list[list[float]] = []
    if geo.get("type") == "LineString":
        coords = geo.get("coordinates") or []
    elif geo.get("type") == "FeatureCollection":
        for f in geo.get("features") or []:
            g = (f or {}).get("geometry") or {}
            if g.get("type") == "LineString":
                coords = g.get("coordinates") or []
                break
    out: list[dict[str, Any]] = []
    for i, c in enumerate(coords):
        if not isinstance(c, (list, tuple)) or len(c) < 2:
            continue
        out.append({"lon": float(c[0]), "lat": float(c[1]), "elapsed_sec": i})
    return out


def _list_has_values(arr: Any) -> bool:
    if not isinstance(arr, list):
        return False
    return any(v is not None for v in arr)


def geojson_has_point_telemetry(props: dict[str, Any]) -> bool:
    """True when properties arrays include usable speed or heart rate."""
    if not isinstance(props.get("elapsed_sec"), list) or not props["elapsed_sec"]:
        return False
    return _list_has_values(props.get("speed_kmh")) or _list_has_values(props.get("heart_rate"))


def _haversine_m(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    earth_r = 6_371_000.0
    to_rad = math.radians
    d_lat = to_rad(lat2 - lat1)
    d_lon = to_rad(lon2 - lon1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) * math.sin(d_lon / 2) ** 2
    )
    return 2 * earth_r * math.asin(math.sqrt(a))


def _segment_speed_kmh(a: dict[str, Any], b: dict[str, Any]) -> float:
    from_data = _float_or_none(a.get("speed_kmh")) or _float_or_none(b.get("speed_kmh"))
    if from_data is not None and from_data > 0:
        return from_data
    dt = float(b["elapsed_sec"]) - float(a["elapsed_sec"])
    if dt <= 0:
        return 0.0
    dist_m = _haversine_m(float(a["lat"]), float(a["lon"]), float(b["lat"]), float(b["lon"]))
    if dist_m <= 0:
        return 0.0
    return (dist_m / dt) * 3.6


def _nearest_row_by_sec(
    rows: list[dict[str, Any]],
    sec: int,
    *,
    sec_key: str,
    max_delta_sec: int = 2,
) -> dict[str, Any]:
    if not rows:
        return {}
    exact = next((r for r in rows if int(r[sec_key]) == sec), None)
    if exact:
        return exact
    best = min(rows, key=lambda r: abs(int(r[sec_key]) - sec))
    if abs(int(best[sec_key]) - sec) <= max_delta_sec:
        return best
    return {}


def derive_point_speeds(points: list[dict[str, Any]]) -> None:
    """Fill missing speed_kmh from GPS segments (in-place)."""
    for i, point in enumerate(points):
        if _float_or_none(point.get("speed_kmh")):
            continue
        for neighbor in (i + 1, i - 1):
            if neighbor < 0 or neighbor >= len(points):
                continue
            a = points[min(i, neighbor)]
            b = points[max(i, neighbor)]
            spd = _segment_speed_kmh(a, b)
            if spd > 0:
                point["speed_kmh"] = spd
                break


def merge_telemetry_into_track_points(
    points: list[dict[str, Any]],
    sensor_rows: list[dict[str, Any]],
    hr_rows: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Normalize track points with sensors, HR, and derived GPS speed."""
    if not points:
        return points
    merged: list[dict[str, Any]] = []
    for point in points:
        sec = int(round(float(point.get("elapsed_sec") or 0)))
        sensor = _nearest_row_by_sec(sensor_rows, sec, sec_key="elapsed_sec", max_delta_sec=2)
        hr = _nearest_row_by_sec(hr_rows, sec, sec_key="seconds", max_delta_sec=30)
        row = dict(point)
        if row.get("speed_kmh") is None and sensor.get("speed_kmh") is not None:
            row["speed_kmh"] = _float_or_none(sensor.get("speed_kmh"))
        if row.get("cadence") is None and sensor.get("cadence") is not None:
            row["cadence"] = _float_or_none(sensor.get("cadence"))
        if row.get("elevation_m") is None and sensor.get("elevation_m") is not None:
            row["elevation_m"] = _float_or_none(sensor.get("elevation_m"))
        if row.get("temperature_c") is None and sensor.get("temperature_c") is not None:
            row["temperature_c"] = _float_or_none(sensor.get("temperature_c"))
        if row.get("power_watts") is None and sensor.get("power_watts") is not None:
            row["power_watts"] = _float_or_none(sensor.get("power_watts"))
        if row.get("heart_rate") is None and hr.get("heart_rate") is not None:
            row["heart_rate"] = int(hr["heart_rate"])
        if row.get("distance_m") is None and hr.get("distance_m") is not None:
            row["distance_m"] = _float_or_none(hr.get("distance_m"))
        merged.append(row)
    derive_point_speeds(merged)
    return merged


def enrich_geojson_from_sensors(
    geo: dict[str, Any],
    sensor_rows: list[dict[str, Any]],
    hr_rows: list[dict[str, Any]],
) -> dict[str, Any]:
    """Дополнить GeoJSON телеметрией из workout_sensors и workout_heart_rate."""
    points = geojson_to_track_points(geo)
    if len(points) < 2:
        return geo
    merged = merge_telemetry_into_track_points(points, sensor_rows, hr_rows)
    enriched = build_enriched_geojson(merged)
    if not enriched:
        return geo
    props = {}
    if geo.get("type") == "FeatureCollection" and geo.get("features"):
        props = (geo["features"][0] or {}).get("properties") or {}
    elif geo.get("type") == "Feature":
        props = geo.get("properties") or {}
    start_time = props.get("start_time")
    if start_time:
        enriched["features"][0]["properties"]["start_time"] = start_time
    return enriched

File: frontend/src/utils/bikeTrack.ts
import {
  resolveSpeedColorProfile,
  speedToColor,
  type SpeedColorProfile,
} from "../config/speedColorScale";

export type { SpeedColorProfile };
/** @deprecated use SpeedColorProfile */
export type CyclingWorkoutProfile = SpeedColorProfile;

export interface TrackPoint {
  lat: number;
  lon: number;
  elapsedSec: number;
  speedKmh?: number | null;
  cadence?: number | null;
  elevationM?: number | null;
  temperatureC?: number | null;
  heartRate?: number | null;
  distanceM?: number | null;
  powerWatts?: number | null;
}

export interface ParsedTrack {
  points: TrackPoint[];
  startTime?: string | null;
}

export interface SpeedSegment {
  positions: [number, number][];
  color: string;
}

function numAt(arr: unknown, i: number): number | null {
  if (!Array.isArray(arr) || i >= arr.length) return null;
  const v = arr[i];
  if (v == null || v === "") return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

/** Статистика скоростей тренировки (для UI, не для раскраски). */
export interface WorkoutSpeedStats {
  min: number;
  max: number;
  avg: number;
  count: number;
}

export function buildWorkoutSpeedStats(points: TrackPoint[]): WorkoutSpeedStats | null {
  const speeds = points
    .map((p) => p.speedKmh)
    .filter((s): s is number => s != null && s > 0);
  if (speeds.length === 0) return null;
  const sum = speeds.reduce((a, b) => a + b, 0);
  return {
    min: Math.min(...speeds),
    max: Math.max(...speeds),
    avg: sum / speeds.length,
    count: speeds.length,
  };
}

/** GeoJSON FeatureCollection с properties-массивами или legacy LineString. */
export function parseTrackGeojson(geo: Record<string, unknown>): ParsedTrack {
  let feature: Record<string, unknown> | null = null;
  if (geo.type === "FeatureCollection" && Array.isArray(geo.features) && geo.features[0]) {
    feature = geo.features[0] as Record<string, unknown>;
  } else if (geo.type === "Feature") {
    feature = geo;
  } else if (geo.type === "LineString") {
    feature = { geometry: geo, properties: {} };
  }

  const geometry = (feature?.geometry ?? geo.geometry) as Record<string, unknown> | undefined;
  const props = (feature?.properties ?? {}) as Record<string, unknown>;
  const coords = (geometry?.coordinates ?? geo.coordinates) as unknown;

  const lines: number[][][] = [];
  if (geometry?.type === "LineString" && Array.isArray(coords)) {
    lines.push(coords as number[][]);
  } else if (geometry?.type === "MultiLineString" && Array.isArray(coords)) {
    for (const line of coords as number[][][]) {
      lines.push(line);
    }
  }

  const flat: number[][] = lines.flat();
  const elapsed = props.elapsed_sec as unknown;
  const hasProps = Array.isArray(elapsed) && elapsed.length > 0;

  const points: TrackPoint[] = [];
  for (let i = 0; i < flat.length; i++) {
    const c = flat[i];
    if (!Array.isArray(c) || c.length < 2) continue;
    const lon = Number(c[0]);
    const lat = Number(c[1]);
    if (!Number.isFinite(lat) || !Number.isFinite(lon)) continue;

    points.push({
      lon,
      lat,
      elapsedSec: hasProps ? (numAt(elapsed, i) ?? i) : i,
      speedKmh: numAt(props.speed_kmh, i),
      cadence: numAt(props.cadence, i),
      elevationM: numAt(props.elevation_m, i),
      temperatureC: numAt(props.temperature_c, i),
      heartRate: numAt(props.heart_rate, i) != null ? Math.round(numAt(props.heart_rate, i)!) : null,
      distanceM: numAt(props.distance_m, i),
      powerWatts: numAt(props.power_watts, i),
    });
  }

  return { points, startTime: (props.start_time as string) ?? null };
}

/** Fill missing speed from GPS segments; idempotent when telemetry already present. */
export function enrichTrackPoints(points: TrackPoint[]): TrackPoint[] {
  if (points.length < 2) return points;
  const out = points.map((p) => ({ ...p }));
  for (let i = 0; i < out.length; i += 1) {
    const point = out[i];
    if (point.speedKmh != null && point.speedKmh > 0) continue;
    for (const neighbor of [i + 1, i - 1]) {
      if (neighbor < 0 || neighbor >= out.length) continue;
      const a = out[Math.min(i, neighbor)];
      const b = out[Math.max(i, neighbor)];
      const spd = segmentSpeedKmh(a, b);
      if (spd > 0) {
        point.speedKmh = spd;
        break;
      }
    }
  }
  return out;
}

export function formatElapsed(sec: number): string {
  const h = Math.floor(sec / 3600);
  const m = Math.floor((sec % 3600) / 60);
  const s = Math.floor(sec % 60);
  if (h > 0) return `${h}:${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")}`;
  return `${m}:${String(s).padStart(2, "0")}`;
}

export function formatClockFromStart(startIso: string | null | undefined, elapsedSec: number): string {
  if (!startIso) return formatElapsed(elapsedSec);
  const base = new Date(startIso.replace(" ", "T"));
  if (Number.isNaN(base.getTime())) return formatElapsed(elapsedSec);
  const t = new Date(base.getTime() + elapsedSec * 1000);
  return t.toLocaleTimeString("ru-RU", { hour: "2-digit", minute: "2-digit", second: "2-digit" });
}

export function formatDistance(m: number | null | undefined): string {
  if (m == null || !Number.isFinite(m)) return "—";
  if (m >= 1000) return `${(m / 1000).toFixed(2)} км`;
  return `${Math.round(m)} м`;
}

const EARTH_R = 6371000;

function haversineM(a: TrackPoint, b: TrackPoint): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLon = toRad(b.lon - a.lon);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_R * Math.asin(Math.sqrt(h));
}

/** Скорость на отрезке: из данных или GPS+время. */
export function segmentSpeedKmh(a: TrackPoint, b: TrackPoint): number {
  const fromData = a.speedKmh ?? b.speedKmh;
  if (fromData != null && fromData > 0) return fromData;

  const dt = b.elapsedSec - a.elapsedSec;
  if (dt > 0) {
    const distM = haversineM(a, b);
    if (distM > 0) return (distM / dt) * 3.6;
  }
  return 0;
}

function mergeAdjacentSegments(segments: SpeedSegment[]): SpeedSegment[] {
  if (segments.length <= 1) return segments;
  const out: SpeedSegment[] = [{ ...segments[0], positions: [...segments[0].positions] }];
  for (let i = 1; i < segments.length; i += 1) {
    const seg = segments[i];
    const last = out[out.length - 1];
    if (last.color === seg.color) {
      last.positions.push(seg.positions[seg.positions.length - 1]);
    } else {
      out.push({ color: seg.color, positions: [...seg.positions] });
    }
  }
  return out;
}

/**
 * Сегменты маршрута с абсолютной шкалой скорости (км/ч).
 * Каждое ребро сохраняется; соседние сегменты одного цвета объединяются.
 */
export function buildSpeedSegments(
  points: TrackPoint[],
  options?: { mergeColors?: boolean; speedProfile?: SpeedColorProfile; workoutType?: string | null },
): SpeedSegment[] {
  if (points.length < 2) return [];

  const profile = options?.speedProfile ?? resolveSpeedColorProfile(options?.workoutType);
  let hasAnySpeed = false;

  const raw: SpeedSegment[] = [];
  for (let i = 0; i < points.length - 1; i += 1) {
    const a = points[i];
    const b = points[i + 1];
    const spd = segmentSpeedKmh(a, b);
    if (spd > 0) hasAnySpeed = true;
    const color = speedToColor(spd, profile);
    raw.push({
      positions: [
        [a.lat, a.lon],
        [b.lat, b.lon],
      ],
      color,
    });
  }

  if (!hasAnySpeed) {
    return [
      {
        positions: points.map((p) => [p.lat, p.lon] as [number, number]),
        color: "#059669",
      },
    ];
  }

  return options?.mergeColors === false ? raw : mergeAdjacentSegments(raw);
}

/** Re-export для centralized color mapping */
export { speedToColor, resolveSpeedColorProfile } from "../config/speedColorScale";

export function findNearestPoint(
  points: TrackPoint[],
  lat: number,
  lon: number,
  maxDistDeg = Infinity,
): TrackPoint | null {
  if (!points.length) return null;
  const maxD2 = maxDistDeg ** 2;
  let best: TrackPoint | null = null;
  let bestD = maxD2;

  const stride = points.length > 3000 ? Math.ceil(points.length / 800) : 1;
  for (let i = 0; i < points.length; i += stride) {
    const p = points[i];
    const d = (p.lat - lat) ** 2 + (p.lon - lon) ** 2;
    if (d < bestD) {
      bestD = d;
      best = p;
    }
  }

  if (best && stride > 1) {
    const idx = points.indexOf(best);
    const from = Math.max(0, idx - stride);
    const to = Math.min(points.length - 1, idx + stride);
    for (let i = from; i <= to; i += 1) {
      const p = points[i];
      const d = (p.lat - lat) ** 2 + (p.lon - lon) ** 2;
      if (d < bestD) {
        bestD = d;
        best = p;
      }
    }
  }

  return bestD < maxD2 ? best : null;
}

const ELEVATION_DISTANCE_STEP_M = 25;

/** Прореживание высоты для графика «высота по дистанции»: ~1 точка на 25 м. */
export function downsampleElevationByDistance(
  elapsedSec: number[],
  distanceM: (number | null)[],
  elevationM: (number | null)[],
  stepM = ELEVATION_DISTANCE_STEP_M,
): { xKm: number[]; y: number[]; elapsedSec: number[] } {
  const samples: { distM: number; elev: number; elapsed: number }[] = [];
  for (let i = 0; i < elapsedSec.length; i++) {
    const d = distanceM[i];
    const e = elevationM[i];
    if (d == null || e == null || !Number.isFinite(d) || !Number.isFinite(e)) continue;
    samples.push({ distM: d, elev: e, elapsed: elapsedSec[i] });
  }
  if (samples.length === 0) {
    return { xKm: [], y: [], elapsedSec: [] };
  }

  samples.sort((a, b) => a.distM - b.distM);

  const smoothWindow = 5;
  const smoothed = samples.map((_, i) => {
    const start = Math.max(0, i - Math.floor(smoothWindow / 2));
    const end = Math.min(samples.length, i + Math.ceil(smoothWindow / 2));
    let sum = 0;
    let n = 0;
    for (let j = start; j < end; j++) {
      sum += samples[j].elev;
      n += 1;
    }
    return { ...samples[i], elev: sum / n };
  });

  const out: typeof smoothed = [];
  let lastKeptDist = -Infinity;
  for (const s of smoothed) {
    if (out.length === 0 || s.distM - lastKeptDist >= stepM) {
      out.push(s);
      lastKeptDist = s.distM;
    }
  }
  const last = smoothed[smoothed.length - 1];
  if (out.length === 0 || Math.abs(out[out.length - 1].distM - last.distM) > 0.01) {
    out.push(last);
  }

  return {
    xKm: out.map((p) => p.distM / 1000),
    y: out.map((p) => Math.round(p.elev * 10) / 10),
    elapsedSec: out.map((p) => p.elapsed),
  };
}
